### PR TITLE
Map our diagnostics to rustc and clippy's ones

### PR DESCRIPTION
**Pull request by HKalbasi**
And control their severity by lint attributes `#[allow]`, `#[deny]` and ... .

It doesn't work with proc macros and I would like to fix that before merge but I don't know how to do it.

### CODE REVIEW

**Comment by Veykril** [comment]
Huh, I canceled an old pending review which caused it to be posted? 🤨

File: crates/ide-diagnostics/src/handlers/incorrect_case.rs [hunk 1]
```diff
@@ -1,28 +1,34 @@
...
+        ctx,
+        code,
         format!(
             "{} `{}` should have {} name, e.g. `{}`",
             d.ident_type, d.ident_text, d.expected_case, d.suggested_text
         ),
-        ctx.sema.diagnostics_display_range(InFile::new(d.file, d.ident.clone().into())).range,
+        InFile::new(d.file, d.ident.clone().into()),
     )
-    .severity(Severity::WeakWarning)
```
Comment (Veykril) [inline crates/ide-diagnostics/src/handlers/incorrect_case.rs:25]: Do we properly handle allows for this now in all cases? If not that should stay a weak warning, I recall there being some issues around us not honoring the allows in all positions.

File: crates/ide-diagnostics/src/lib.rs [hunk 2]
```diff
@@ -74,52 +74,105 @@ use ide_db::{
...
-impl DiagnosticCode {
-    pub fn as_str(&self) -> &str {
-        self.0
+impl DiagnosticOrigin {
+    pub fn url(&self) -> String {
+        match self {
+            DiagnosticOrigin::RustcHardError(e) => {
+                format!("https://doc.rust-lang.org/stable/error_codes/{e}.html")
+            }
+            DiagnosticOrigin::RustcLint(e) => format!("https://duckduckgo.com/?q=rustc+lint+{e}"),
```
[the remark this answers is not in the document]
Comment (HKalbasi) [reply crates/ide-diagnostics/src/lib.rs]: I replaced this with a mdbook search in rustc book, is it good? Rustc book contains info for all lints but computing the link for a lint requires knowing its severity by default (example: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unused-mut) which we don't know currently but we should at some point.

File: crates/ide-diagnostics/src/lib.rs [hunk 3]
```diff
@@ -74,52 +74,105 @@ use ide_db::{
...
+    ast::{self, AstNode},
+    SyntaxNode, SyntaxNodePtr, TextRange,
 };
-use syntax::{algo::find_node_at_range, ast::AstNode, SyntaxNodePtr, TextRange};
 
 // FIXME: Make this an enum
-#[derive(Copy, Clone, Debug, PartialEq)]
-pub struct DiagnosticCode(pub &'static str);
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum DiagnosticOrigin {
```
[the remark this answers is not in the document]
Comment (HKalbasi) [reply crates/ide-diagnostics/src/lib.rs:93]: Maybe the old `DiagnosticCode` was good?

**Comment by HKalbasi** [comment]
Let's merge this now to find its potential problems in nightly releases.
@bors r+

**Comment by Veykril** [comment]
oh sorry, I think I missed your ping back then 😅

**Comment by HKalbasi** [comment]
That's fine :) I was just not super confident about the changes in `ide-diagnostics/src/lib.rs`. I can still fix problems if you see any.

**Pull request merged**